### ISSUE TITLE
assertion failed: B > A (in hz_linesearch!)

### ISSUE THREAD

**Issue by oxinabox** [JuliaOpt/Optim.jl#121]
I hit an assert while using 

On [this line](https://github.com/JuliaOpt/Optim.jl/blob/6f0be0643d2795c5e9169e63283c2cc5ab7cd309/src/linesearch/hz_linesearch.jl#L341)

```
res = optimize(f, θ, method=:l_bfgs, show_trace = true, store_trace = true, iterations = 50);
```

`f` is a loss function (and its gradient), for a special type of neural network. I know this type of network has historically been trained with LBFGS, and indeed it normally works quiet well for me. but this time, it does not. 

**Does the assert check part of Optim's logic?**

Or can it be triggered by a pathological enough `f`?

**Comment by timholy** [comment]
The asserts are there to check the linesearch logic (which is quite complicated); they basically force the developer to discover and fix problems. That said, some of the problems come from a broken `f`, and it's not always easy to distinguish which is which.

I'd start by putting `@show a b ia ib A B iA iB lsr` on the line before the assert; look at the last output and see what it tells you about what's going on.

**Comment by timholy** [comment]
I should also add that, historically, most problems have arisen from functions with broken gradients. I can't stress enough how important it is to check your gradients **obsessively**, for example with DualNumbers. A gradient that matches to 1 part in 10^9, by DualNumbers and `Float64` precision, is not "good enough": it's a demonstration that you have a bug in your code. When using DualNumbers, they should match basically to the machine precision (1 part in 10^15 or so). When you encounter a problem, it's also important to check them at the approximate location of the error, since the error can depend on the location.

**Comment by oxinabox** [comment]
I've only checked the gradient using small (real) epsilons.
I'll check it with Dual Numbers now.
(Once I work out how)
Thanks

**Comment by timholy** [comment]
Once you know how, you'll never go back :smile:. It's simply a dream, and one of the coolest mathematical points I've learned via participating in Julia development.

If you need to check Hessians, you can use HyperDualNumbers.

**Comment by johnmyleswhite** [comment]
This issue comes up often enough that I wonder if we should try to do a gradient check automatically whenever the line search fails. It would certainly provide a much more informative error message.

**Comment by timholy** [comment]
Good point. Alternatively, an error message stating the likely problem along with a `check_gradient` function? As upsides of doing it automatically, as you suggest, seem to be: (1) it's easy for the user, and (2) evaluates the derivative at the position (as in `x`) where the error occurred. The potential downsides I see are that (3) a user's function may have declared types like `f!(x::Vector{Float64})` or something, and (4) an `N`-dimensional problem requires `N` function evaluations to check the entire gradient, which could be trouble in some cases. Not quite sure how to balance these considerations. In a sense, even a recommended `check_gradient` function could run into the same problem.

Maybe we could:
- issue a warning: "An unexpected situation has arisen. Optim is checking that your gradient function produces numerically-accurate answers using DualNumbers. If you receive a method error, it indicates that your function's type declaration would benefit from generalization."
- start running the to-be-written function `check_gradient(f, x, timeout=5.0)`, which will test those gradient components that have been evaluated by the time ~5s have elapsed. In some cases this might check only a single component (if each function evaluation takes more than 5s), but we can also issue a warning that checking was incomplete.
- error out with the most likely diagnosis: either the user's gradient is wrong, or optim is just confused.

**Comment by johnmyleswhite** [comment]
One solution to the `f(x::Vector{Float64}` problem is to fall back on finite-differencing if necessary. It obviously can have accuracy issues that might introduce false positives, but my experience is that gradients are usually wildly wrong, not just numerically unstable.

I think the timeout approach along with making automatic gradient checking an option that can be turned off would address our biggest worries while making it easier to provide more informative error messages.

**Comment by pkofod** [comment]
Regarding the point that sometimes automatic derivative checking might cause running DualNumbers on a massive problem (`N` very large). It might also be that a single function evaluation simply takes a very long time. This talks against running it automatically. Couldn't the error just cause the optimizer to return the appropriate `OptimizationResults` object, with `false` conversion flags, and the current `x`. The user can then call `check_gradient(f, results.x)` to see if there were gradient problems. Currently, this is difficult as the optimize call simply quits.

**Comment by timholy** [comment]
I had the same thought. But the problem is, what if the user doesn't checked the `converged` flag and thinks everything is OK?

**Comment by oxinabox** [comment]
throw a `warn`, to indicate everything is not ok.

(Still working on making my code work with dual numbers -- my functions were all defined in terms of Vector/Matrix{Float64})

**Comment by oxinabox** [comment]
Checked with DualNumbers, via ForwardDiff.jl.

I can't check the full exact parameters,
as the loss is a the mean error over 4000 tuples of ~10 vectors of length 300.
And N in my case (the number of parameters) is 10150.

I suspect running ForwardDiff.jl on that might take weeks.

So I checked just two of these tuples, at two random sets of parameters.
Then checked the maximum absolute value of the difference between my analytic  grad and  that I get from dual numbers.
The 4 max differences I found were 4.440892098500626e-16, 1.7763568394002505e-15, 2.6645352591003757e-15  5.551115123125783e-16

So the worse was being "off" by 2.66e-15
That is pretty good, right?

These numbers give me confidance my implementation is correct,
Though does not prove it is correct at the place where the error occurs.

---

@timholy You're right, checking graidents with dual numbers is the coolest thing I've seen since checking backprop gradients via finite difference. (Which was game changing for me, so much less worry about implementation mistakes.)

**Comment by timholy** [comment]
Probably looks good. I'm assuming that either these are relative errors and/or that the absolute value of your objective function is around 1 (an absolute error on the scale of 1e-16 is not small if the value of the function is 1e-12). Presuming that's not an issue here, that's all consistent with normal roundoff error, and thus demonstrates that your gradients are bug-free. Glad you enjoyed the outcome despite its mere confirmation that you're on the right track.

OK, as a next step I'd recommend inserting that `@show` line I posted at the top, and seeing what it says just before the error. The variables that start with `i` are index variables into a "line search history" object `lsr`, which records each evaluation point (location, objective function value, and projected slope) during a line search. The `a`, `b`, `A`, `B` variables represent the actual corresponding locations along the line.

**Comment by oxinabox** [comment]
Indeed the absolute value of my function (before optimising) is >1, normally 3-14 for random initalistations.

I have the show line in place (well I will having in place again shortly, moved back to 4.0).
The assert fail doesn't seem to have reoccurred,
Though it occured several times before.
I guess it is indeed numerically sensitive to particular state.

I suspect from here the issue may sit for days-weeks waiting for that state to reoccur.

**Comment by timholy** [comment]
OK, if you find it and can't figure it out on your own, please do copy/paste the relevant output here.

**Comment by johnmyleswhite** [comment]
Closing this one since we don't have a minimal test case yet and it may be solved by #144.
